test/functional/legacy/061_undo_tree_spec.lua

```diff
@@ -0,0 +1,256 @@
+-- Tests for undo tree and :eralier and :later.
+
+local helpers = require('test.functional.helpers')
+local feed, insert, source, eq, eval, clear, execute, expect, wait =
+  helpers.feed, helpers.insert, helpers.source, helpers.eq, helpers.eval,
+  helpers.clear, helpers.execute, helpers.expect, helpers.wait
+
+local expect_empty_buffer = function()
+  -- The space will be removed by helpers.dedent but is needed as dedent will
+  -- throw an error if it can not find the common indent of the given lines.
+  return expect(' ')
+end
+local expect_line = function(line)
+  return eq(line, eval('getline(".")'))
+end
+
+describe('the undo tree', function()
+  setup(function()
+    clear()
+    local file = io.open('Xtest.source', 'w')
+    file:write('o1\x1ba2\x12=string(123)\n\x1b')
+    file:flush()
+    file:close()
+  end)
+  teardown(function()
+    os.remove('Xtest')
+    os.remove('Xtest.source')
+  end)
+
+  it('is working', function()
+    -- Assert that no undo history is present.
+    eq({}, eval('undotree().entries'))
+    insert([[
+      
+      123456789]])
+
+    -- Clear the undo history after the insertion (see :h clear-undo)
+    execute('let old_undolevels = &undolevels')
+    execute('set undolevels=-1')
+    feed('a <BS><Esc>') 
+    execute('let &undolevels = old_undolevels')
+    execute('unlet old_undolevels')
+    eq({}, eval('undotree().entries'))
+
+    -- Delete three characters and undo.
+    feed('Gxxx')
+    expect_line('456789')
+    feed('g-')
+    expect_line('3456789')
+    feed('g-')
+    expect_line('23456789')
+    feed('g-')
+    expect_line('123456789')
+    feed('g-')
+    expect_line('123456789')
+
+    -- Delete three other characters and go back in time step by step.
+    feed('$xxx')
+    expect_line('123456')
+    execute('sleep 1')
+    wait()
+    feed('g-')
+    expect_line('1234567')
+    feed('g-')
+    expect_line('12345678')
+    feed('g-')
+    expect_line('456789')
+    feed('g-')
+    expect_line('3456789')
+    feed('g-')
+    expect_line('23456789')
+    feed('g-')
+    expect_line('123456789')
+    feed('g-')
+    expect_line('123456789')
+    feed('g-')
+    expect_line('123456789')
+    feed('10g+')
+    expect_line('123456')
+
+    -- Delay for three seconds and go some seconds forward and backward.
+    execute('sleep 2')
+    wait()
+    feed('Aa<esc>')
+    feed('Ab<esc>')
+    feed('Ac<esc>')
+    expect_line('123456abc')
+    execute('ear 1s')
```

Comment:
I'm just guessing, but maybe add a `wait()` call after these `:earlier` commands to see if that helps with the random test failures. 

Also, I suggest spelling out the full command (`:earlier` rather than `:ear`) unless there's some space constraints that make it onerous to do so.